base/iobuffer.jl

```diff
@@ -60,57 +72,45 @@ julia> String(take!(io))
 "JuliaLang is a GitHub organization"
 ```
 """
-IOBuffer(data::AbstractVector{UInt8}, readable::Bool=true, writable::Bool=false, maxsize::Integer=typemax(Int)) =
-    GenericIOBuffer(data, readable, writable, true, false, maxsize)
-function IOBuffer(readable::Bool, writable::Bool)
-    b = IOBuffer(StringVector(32), readable, writable)
-    b.data[:] = 0
-    b.size = 0
-    return b
+function IOBuffer(
+        data::AbstractVector{UInt8};
+        read::Union{Bool,Nothing}=nothing,
+        write::Union{Bool,Nothing}=nothing,
+        append::Union{Bool,Nothing}=nothing,
+        truncate::Union{Bool,Nothing}=nothing,
+        maxsize::Integer=typemax(Int))
+    if maxsize < 0
+        throw(ArgumentError("negative maxsize: $(maxsize)"))
+    end
+    flags = open_flags(read=read, write=write, append=append, truncate=truncate)
+    buf = GenericIOBuffer(data, flags.read, flags.write, true, flags.append, Int(maxsize))
+    if flags.truncate
+        buf.size = 0
+    end
+    return buf
 end
 
-"""
-    IOBuffer() -> IOBuffer
-
-Create an in-memory I/O stream, which is both readable and writable.
-
-# Examples
-```jldoctest
-julia> io = IOBuffer();
-
-julia> write(io, "JuliaLang is a GitHub organization.", " It has many members.")
-56
-
-julia> String(take!(io))
-"JuliaLang is a GitHub organization. It has many members."
-```
-"""
-IOBuffer() = IOBuffer(true, true)
-
-"""
-    IOBuffer(size::Integer)
-
-Create a fixed size IOBuffer. The buffer will not grow dynamically.
-
-# Examples
-```jldoctest
-julia> io = IOBuffer(12)
-IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=12, ptr=1, mark=-1)
-
-julia> write(io, "Hello world.")
-12
-
-julia> String(take!(io))
-"Hello world."
-
-julia> write(io, "Hello world again.")
-12
-
-julia> String(take!(io))
-"Hello world "
-```
-"""
-IOBuffer(maxsize::Integer) = (x=IOBuffer(StringVector(maxsize), true, true, maxsize); x.size=0; x)
+function IOBuffer(;
+        read::Union{Bool,Nothing}=true,
+        write::Union{Bool,Nothing}=true,
+        append::Union{Bool,Nothing}=nothing,
+        truncate::Union{Bool,Nothing}=true,
+        maxsize::Integer=typemax(Int))
+    size = maxsize == typemax(Int) ? 32 : Int(maxsize)
+    flags = open_flags(read=read, write=write, append=append, truncate=truncate)
+    buf = IOBuffer(
+        StringVector(size),
+        read=flags.read,
+        write=flags.write,
+        append=flags.append,
+        truncate=flags.truncate,
+        maxsize=maxsize)
+    buf.data[:] = 0
+    if flags.truncate
+        buf.size = 0
```

Comment:
Not important, just wondering: since `truncate` is also passed to the constructor above, is this redundant?

Reply:
Yes, it is. Fixed in #25919.